From 467cf3ca07b3eb1f2341dd5ba5ba1081fac55968 Mon Sep 17 00:00:00 2001
From: Robert Mader <robert.mader@collabora.com>
Date: Sun, 28 Sep 2025 17:08:34 +0200
Subject: [PATCH 3/3] aperture: Add image encoder properties map

Improve the jpeg quality over the very performance oriented jpegenc
defaults.

(cherry picked from commit 41c50e2b27f17a05b25f106cb93359cc4f030b19)
---
 aperture/src/viewfinder.rs | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/aperture/src/viewfinder.rs b/aperture/src/viewfinder.rs
index bf998e8..bfeb0e0 100644
--- a/aperture/src/viewfinder.rs
+++ b/aperture/src/viewfinder.rs
@@ -948,6 +948,16 @@ impl Viewfinder {
             )
             .build();
 
+        let image_properties_map = ElementProperties::builder_map()
+            .item(
+                ElementPropertiesMapItem::builder("jpegenc")
+                    .field("quality", 95)
+                    // idct-method "float": Slowest, most accurate method.
+                    .field("idct-method", 2)
+                    .build(),
+            )
+            .build();
+
         let video_profile = match self.video_format() {
             VideoFormat::H264Mp4 => {
                 let mut hw_encoder_found = false;
@@ -1070,8 +1080,15 @@ impl Viewfinder {
             }
         };
 
+        let image_profile =
+            gst_pbutils::EncodingVideoProfile::builder(&gst::Caps::builder("image/jpeg").build())
+                .variable_framerate(true)
+                .element_properties(image_properties_map)
+                .build();
+
         let camerabin = self.imp().camerabin();
         camerabin.set_property("video-profile", video_profile);
+        camerabin.set_property("image-profile", image_profile);
     }
 
     fn init(&self) {
-- 
2.51.0

